### PR TITLE
Move seeking logic into a custom editor clock

### PR DESCRIPTION
**Pull request by smoogipoo**
* `Editor` now has an `EditorClock`.
* Moves the seek handling from `HitObjectComposer` to `EditorClock`.
* Cleans up a lot of testcases by encapsulating the clock construction/handling logic.
* Exposed as an `IAdjustableClock` to editor components through DI - if we need components to snap later on then we may expose it as the `EditorClock` itself, but as we discussed earlier this is not needed for now.

### CODE REVIEW

File: osu.Game/Tests/Visual/EditorClockTestCase.cs [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+            Clock.ControlPointInfo = working.Beatmap.ControlPointInfo;
+            Clock.ChangeSource((IAdjustableClock)working.Track ?? new StopwatchClock());
+            Clock.ProcessFrame();
+        }
+
+        protected override void Update()
+        {
+            base.Update();
+
+            // We don't have any explicit way to start/stop the track, but want a relatively accurate IsRunning state
```
Comment (peppy) [inline osu.Game/Tests/Visual/EditorClockTestCase.cs]: i'm not sure this comment explains what's going on here so well... i can't understand it.

File: osu.Game/Screens/Edit/Components/Timelines/Summary/SummaryTimeline.cs [hunk 2]
```diff
@@ -17,8 +17,15 @@ namespace osu.Game.Screens.Edit.Components.Timelines.Summary
     /// </summary>
     public class SummaryTimeline : BottomBarContainer
     {
+        private readonly IAdjustableClock adjustableClock;
+
+        public SummaryTimeline(IAdjustableClock adjustableClock)
```
Comment (peppy) [inline osu.Game/Screens/Edit/Components/Timelines/Summary/SummaryTimeline.cs]: what's the reasoning behind moving this out of DI?

File: osu.Game/Screens/Edit/EditorClock.cs [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+        private readonly BindableBeatDivisor beatDivisor;
+
+        public EditorClock(ControlPointInfo controlPointInfo, BindableBeatDivisor beatDivisor)
+        {
+            this.beatDivisor = beatDivisor;
+
+            ControlPointInfo = controlPointInfo;
+        }
+
+        public bool SeekSnapped(double position)
```
Comment (peppy) [inline osu.Game/Screens/Edit/EditorClock.cs]: probably best to xmldoc this (and hopefully this class too)

File: osu.Game/Screens/Edit/EditorClock.cs [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+
+        public EditorClock(ControlPointInfo controlPointInfo, BindableBeatDivisor beatDivisor)
+        {
+            this.beatDivisor = beatDivisor;
+
+            ControlPointInfo = controlPointInfo;
+        }
+
+        /// <summary>
+        /// Seek to the closest valid snap value.
```
Comment (smoogipoo) [inline osu.Game/Screens/Edit/EditorClock.cs]: This should be something more like "Seek to the closest beat at a time."

File: osu.Game/Screens/Edit/EditorClock.cs [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+            // the next timing point's start time
+            var nextTimingPoint = ControlPointInfo.TimingPoints.FirstOrDefault(t => t.Time > timingPoint.Time);
+            if (position > nextTimingPoint?.Time)
+                position = nextTimingPoint.Time;
+
+            return Seek(position);
+        }
+
+        /// <summary>
+        /// Seeks the current time one beat-snapped beat-length backwards.
```
Comment (smoogipoo) [inline osu.Game/Screens/Edit/EditorClock.cs]: Not necessarily snapped (see param).

File: osu.Game/Screens/Edit/EditorClock.cs [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+        /// </summary>
+        /// <param name="snapped">Whether to snap to the closest beat.</param>
+        public void SeekBackward(bool snapped = false) => seek(-1, snapped);
+
+        /// <summary>
+        /// Seeks the current time one beat-snapped beat-length forwards.
```
Comment (smoogipoo) [inline osu.Game/Screens/Edit/EditorClock.cs]: Not necessarily snapped (see param).

**Review by peppy — APPROVED** [review]

**Pull request merged**